dialog/file.go

```diff
@@ -245,39 +338,85 @@ func (f *fileDialog) setSelected(file *fileDialogItem) {
 		f.selected.Refresh()
 	}
 	if file != nil && file.isDirectory() {
-		f.setDirectory(file.path)
+		lister, err := storage.ListerForURI(file.location)
+		if err != nil {
+			fyne.LogError("Failed to create lister for URI"+file.location.String(), err)
+		}
+		f.setDirectory(lister)
 		return
 	}
 	f.selected = file
 
-	if file == nil || file.path == "" {
+	if file == nil || file.location.String()[len(file.location.Scheme())+3:] == "" {
 		f.fileName.SetText("")
 		f.open.Disable()
 	} else {
 		file.isCurrent = true
-		f.fileName.SetText(filepath.Base(file.path))
+		f.fileName.SetText(file.location.Name())
 		f.open.Enable()
 	}
 }
 
-// effectiveStartingDir calculates the directory at which the file dialog
-// should open, based on the values of  CWD, home, and any error conditions
-// which occur.
+// effectiveStartingDir calculates the directory at which the file dialog should
+// open, based on the values of StartingDirectory, CWD, home, and any error
+// conditions which occur.
 //
 // Order of precedence is:
 //
+// * file.StartingDirectory if non-empty, os.Stat()-able, and uses the file://
+//   URI scheme
+// * os.Getwd()
 // * os.UserHomeDir()
 // * "/" (should be filesystem root on all supported platforms)
-func (f *FileDialog) effectiveStartingDir() string {
+//
+func (f *FileDialog) effectiveStartingDir() fyne.ListableURI {
+	var startdir fyne.ListableURI = nil
+
+	if f.StartingLocation != nil {
+		startdir = f.StartingLocation
+	}
+
+	if startdir != nil {
+		if startdir.Scheme() == "file" {
+			path := startdir.String()[len(startdir.Scheme())+3:]
+
+			// the starting directory is set explicitly
+			if _, err := os.Stat(path); err != nil {
+				fyne.LogError("Error with StartingLocation", err)
+			} else {
+				return startdir
+			}
+		}
+
+	}
+
+	// Try to get ./
+	wd, err := os.Getwd()
```

Comment:
My view is still that HomeDir() should be preferred over PWD. I think that this also was what we agreed on when talking about this in earlier discussions because PWD makes more sense for the developer than what it does for a user. A user of the application would probably be more happy to have Home than PWD.

Reply:
Just a comment from a user.. When an app is launched from an icon, it isn't obvious what PWD would be. This may be as simple as it being documented somewhere but being able to rely on $HOME would be straight forward.

Reply:
My proposal would be to have HomeDir the default and then developers wanting PWD could just put in a temporary different starting directory now that we, with this PR, have support for changing starting directory.

Reply:
This was a debate we had previously in #1111, at which point we decided that ./ would be the best starting location.

Reply:
We discussed this at today’s call and we (@stuartmscott, @andydotxyz and me) concluded that HomeDir should be the default as it is “the least likely to surprise the user” and having PWD can easily be toggled with few lines of code now that this work has landed.

Reply:
Well if that was the decision.

Honestly, I think there has been enough contention on this point that maybe we should consider making it a preference the user can set? I find it infuriating that a few programs I use regularly open `~` rather than my working directory, since I then have to nagivate through many levels of file hierarchy.

A compromise might be to add `./` to "favorites"?

Reply:
Yeah. I think adding PWD to favorites if PWD isn’t HomeDir would be a good addition.